sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -658,7 +658,8 @@ object CollapseProject extends Rule[LogicalPlan] {
 
   def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
```

Comment:
Is the purpose to give up this rule basically?  Why don't we consider using `spark.sql.optimizer.excludedRules`? I think it's more general way to resolve such issues.

Reply:
The purpose is to improve the rule so that it only applies when it will yield a performance improvement (and not apply when it could cause memory issues).  This was the preferred solution, since if we excluded the rule entirely we wouldn't benefit from it in the instances where it would be beneficial.

Reply:
So, basically what you want to do is to give up a rule given a condition because processing huge tree causes OOM issue only in the driver. Am I correct?

What's the diff if we set the threshold `spark.sql.maxRepeatedAliasSize` to set the specific number based upon the rough estimation, and explicitly exclude the rule by `spark.sql.optimizer.excludedRules` based on user's rough estimation?

Reply:
@HyukjinKwon it's not that processing a huge tree causes an OOM, it's that the user can write a small tree, that seems very reasonable to execute, but under the hood the optimiser turns it into a huge tree that OOMs.  The user doesn't know beforehand that the optimiser issue is going to happen, in order to disable the rule.  It takes a lot of debugging, looking through stack traces, etc, to identify that the OOM is caused by CollapseProject and that you can disable it.   Also, we typically run many different queries within a spark session, and wouldn't want to disable CollapseProject for all of them.

This change means that we can still run CollapseProject, we just don't substitute overly large aliases.  In the types of query we had problems with, this means that it will collapse the query until the aliases get too large, and then stop.  So we still do apply CollapseProject to every query, we just stop substituting any alias the gets too large.

`spark.sql.maxRepeatedAliasSize` just determines the size of alias tree that is determined to be too large to efficiently substitute multiple times. The default value of `100` was determined by some basic testing to find the best perf balance (see charts at top), but happy to tweak this if you don't htink it's appropriate?